### PR TITLE
feat: support `pre` caveat in `store/list` and `upload/list` 

### PR DESCRIPTION
**Pull request by travis**
Implement support for the `pre` caveat introduced in https://github.com/web3-storage/w3protocol/pull/423

Also introduce `startCursor` and `endCursor` with the intention to deprecate `cursor`. By providing cursors for the beginning and end of the range, we enable clients to page forward or backward without saving state anywhere. 

Example usage:

Given uploads like:

```bash
$ w3 ls
bafybeia7tr4dgyln7zeyyyzmkppkcts6azdssykuluwzmmswysieyadcbm
bafybeibvbxjeodaa6hdqlgbwmv4qzdp3bxnwdoukay4dpl7aemkiwc2eje
bafybeid2fbwlsb7euczxw7ig3q3fyezw7tfkcikrer75w4ucfawh5vqfdm
bafybeidibxjgswcu2q4gx7j5mzq26qtcczt3nczg6k4ybqct2rm567kpoy
bafybeie7vavzb2ojufppuaxer3kfu4yxi36z7ozdz7wnhtylnd7rl5p6ji
bafybeiefiddap3p3z3p3obatviipgda2us3umt4uxeh2trrni5pt6bfc6e
bafybeifohbynkellbberk3hi447dntungwgicmsvxuwdttk5k5k535mw2a
bafybeigkeedelu7fkqgeafzh7zrtwmcn747clncxuufl2dzu46mww735qq
bafybeign7hb7ynhc6o3sevas365xuzuqzrhpfzgksi4dzpcbphl4xrerye
bafybeignyyeaqn6b2fzqswrozt2txexjbwesrt4dbxlquoac4nn7mbi2gm
bafybeigz7o2ibtv6k5pnklfza3ebsl6jeveguoyxwvrmmtvp2lkvqqjyem
```

You can now page forward like:

```bash
$ w3 can upload ls --size 4
bafybeia7tr4dgyln7zeyyyzmkppkcts6azdssykuluwzmmswysieyadcbm
bafybeibvbxjeodaa6hdqlgbwmv4qzdp3bxnwdoukay4dpl7aemkiwc2eje
bafybeid2fbwlsb7euczxw7ig3q3fyezw7tfkcikrer75w4ucfawh5vqfdm
bafybeidibxjgswcu2q4gx7j5mzq26qtcczt3nczg6k4ybqct2rm567kpoy
```

```bash
$ w3 can upload ls --size 4 --cursor bafybeidibxjgswcu2q4gx7j5mzq26qtcczt3nczg6k4ybqct2rm567kpoy
bafybeie7vavzb2ojufppuaxer3kfu4yxi36z7ozdz7wnhtylnd7rl5p6ji
bafybeiefiddap3p3z3p3obatviipgda2us3umt4uxeh2trrni5pt6bfc6e
bafybeifohbynkellbberk3hi447dntungwgicmsvxuwdttk5k5k535mw2a
bafybeigkeedelu7fkqgeafzh7zrtwmcn747clncxuufl2dzu46mww735qq
```

```bash
$ w3 can upload ls --size 4 --cursor bafybeigkeedelu7fkqgeafzh7zrtwmcn747clncxuufl2dzu46mww735qq
bafybeign7hb7ynhc6o3sevas365xuzuqzrhpfzgksi4dzpcbphl4xrerye
bafybeignyyeaqn6b2fzqswrozt2txexjbwesrt4dbxlquoac4nn7mbi2gm
bafybeigz7o2ibtv6k5pnklfza3ebsl6jeveguoyxwvrmmtvp2lkvqqjyem
```

and page backward like:

```bash
$ w3 can upload ls --size 4 --cursor bafybeign7hb7ynhc6o3sevas365xuzuqzrhpfzgksi4dzpcbphl4xrerye --pre
bafybeigkeedelu7fkqgeafzh7zrtwmcn747clncxuufl2dzu46mww735qq
bafybeifohbynkellbberk3hi447dntungwgicmsvxuwdttk5k5k535mw2a
bafybeiefiddap3p3z3p3obatviipgda2us3umt4uxeh2trrni5pt6bfc6e
bafybeie7vavzb2ojufppuaxer3kfu4yxi36z7ozdz7wnhtylnd7rl5p6ji
```

```bash
$ w3 can upload ls --size 4 --cursor bafybeie7vavzb2ojufppuaxer3kfu4yxi36z7ozdz7wnhtylnd7rl5p6ji --pre
bafybeidibxjgswcu2q4gx7j5mzq26qtcczt3nczg6k4ybqct2rm567kpoy
bafybeid2fbwlsb7euczxw7ig3q3fyezw7tfkcikrer75w4ucfawh5vqfdm
bafybeibvbxjeodaa6hdqlgbwmv4qzdp3bxnwdoukay4dpl7aemkiwc2eje
bafybeia7tr4dgyln7zeyyyzmkppkcts6azdssykuluwzmmswysieyadcbm
```

Note that the cursor values here happen to match the CID strings - this is an implementation detail and should not be relied upon. Instead, cursor values should be taken from `startCursor` and `endCursor` in the list response:

```bash
$ w3 can upload ls --size 4 --cursor bafybeidibxjgswcu2q4gx7j5mzq26qtcczt3nczg6k4ybqct2rm567kpoy --raw
{"size":4,"cursor":"bafybeigkeedelu7fkqgeafzh7zrtwmcn747clncxuufl2dzu46mww735qq","results":[{"root":{"/":"bafybeie7vavzb2ojufppuaxer3kfu4yxi36z7ozdz7wnhtylnd7rl5p6ji"},"shards":[{"/":"bagbaieracmkgwrw6rowsk5jse5eihyhszyrq5w23aqosajyckn2tfbotdcqq"}],"updatedAt":"2023-02-13T16:30:04.172Z","insertedAt":"2023-02-13T16:30:04.172Z"},{"root":{"/":"bafybeiefiddap3p3z3p3obatviipgda2us3umt4uxeh2trrni5pt6bfc6e"},"shards":[{"/":"bagbaieraz3iyicghnyzpjld7yxf7rmntf5cdbewo4pmuck6kkkc6f7z3sznq"}],"updatedAt":"2023-02-13T16:29:51.864Z","insertedAt":"2023-02-13T10:57:39.017Z"},{"root":{"/":"bafybeifohbynkellbberk3hi447dntungwgicmsvxuwdttk5k5k535mw2a"},"shards":[{"/":"bagbaierasu5cufqagk34dkmhszbwuhiydjjjp3pl4bacyvskljqtkcxhgbta"}],"updatedAt":"2023-02-13T16:30:11.648Z","insertedAt":"2023-02-13T10:57:16.264Z"},{"root":{"/":"bafybeigkeedelu7fkqgeafzh7zrtwmcn747clncxuufl2dzu46mww735qq"},"shards":[{"/":"bagbaieralqhbrq2mqupqijkoylagbams263jwq6helsl6kbdpivupjm7egxq"}],"updatedAt":"2023-02-13T16:30:18.403Z","insertedAt":"2023-02-13T10:57:49.695Z"}],
 "endCursor":"bafybeigkeedelu7fkqgeafzh7zrtwmcn747clncxuufl2dzu46mww735qq","startCursor":"bafybeie7vavzb2ojufppuaxer3kfu4yxi36z7ozdz7wnhtylnd7rl5p6ji"}
```

nb that these w3 commands require the changes I implemented in https://github.com/web3-storage/w3cli/pull/45

### CODE REVIEW

File: upload-api/service/types.ts [hunk 1]
```diff
@@ -116,11 +116,14 @@ export interface UploadListItem extends UploadAddResult {
 
 export interface ListOptions {
   size?: number,
-  cursor?: string
+  cursor?: string,
+  pre?: boolean
 }
 
 export interface ListResponse<R> {
   cursor?: string,
```
Comment (vasco-santos) [inline upload-api/service/types.ts:124]: Can we add a comment here that this is being deprecated?

File: upload-api/tables/store.js [hunk 2]
```diff
@@ -119,20 +119,25 @@ export function createStoreTable (region, tableName, options = {}) {
...
         AttributesToGet: ['link', 'size', 'origin', 'insertedAt'],
       })
       const response = await dynamoDb.send(cmd)
 
       const results = response.Items?.map(i => toStoreListResult(unmarshall(i))) ?? []
+      const startCursor = results[0] ? results[0].link.toString() : undefined
       // Get cursor of the item where list operation stopped (inclusive).
       // This value can be used to start a new operation to continue listing.
       const lastKey = response.LastEvaluatedKey && unmarshall(response.LastEvaluatedKey)
       const cursor = lastKey ? lastKey.link : undefined
```
Comment (vasco-santos) [inline upload-api/tables/store.js:133]: should we call this endCursor, and have the return object to have `cursor: endCursor` also with comment that this is going to be deprecated?

**Review by vasco-santos — CHANGES_REQUESTED** [review]
Code changes look good! Thanks @travis 
Can we please add some tests for this under `upload-api/test/service/{store, upload}.test.js`?

**Comment by vasco-santos** [comment]
Also @travis thanks for dealing with all this deps upgrades in the PR ❤️

**Comment by travis** [comment]
> Can we please add some tests for this under `upload-api/test/service/{store, upload}.test.js`?

Ah yes - will jump on this tomorrow most likely!

**Comment by travis** [comment]
ok @vasco-santos got some tests in, please take a look!

two questions came out of this:
a) is it worth trying to get the tests running on my local or will that require extensive AWS config that's more trouble than it's worth?
b) these tests reuse a ton of boilerplate - I see some opportunities to DRY things up quite a bit, but notice you haven't done much of that up until now - are you intentionally avoiding getting rid of some of the boilerplate? if not, any objection to me spending a little time making things a little tighter? I'm thinking that wrapping all of the capability creation in a more readable test-specific API would be pretty straightforward and clean a bunch of these tests up...

**Review by vasco-santos — APPROVED** [review]
LGTM 🚀

**Comment by vasco-santos** [comment]
We synced offline in a, should be good to go with Docker installed locally. But maybe there is a requirement so have some environment variables AWS specific setup (per CI file).

b) we can do in follow up per discussion in sync later

**Pull request merged**